core/app/models/spree/address.rb

```diff
@@ -8,15 +8,31 @@ module Spree
   class Address < Spree::Base
     extend ActiveModel::ForbiddenAttributesProtection
 
+    mattr_accessor :state_validator_class,
+                   :state_normalizer_class
+    self.state_validator_class = Spree::Address::StateValidator
+    self.state_normalizer_class = Spree::Address::StateNormalizer
+
     belongs_to :country, class_name: "Spree::Country", optional: true
     belongs_to :state, class_name: "Spree::State", optional: true
 
     validates :firstname, :address1, :city, :country_id, presence: true
     validates :zipcode, presence: true, if: :require_zipcode?
     validates :phone, presence: true, if: :require_phone?
 
-    validate :state_validate
-    validate :validate_state_matches_country
+    validate do
+      self.class.state_normalizer_class.new(self).perform
+      self.class.state_validator_class.new(self).perform
+
+      %w[state_validate validate_state_matches_country].each do |method|
+        if self.class.method_defined?(method)
+          Rails.logger.warn(
```

Comment:
What if we called the custom methods instead of the new behavior in case the methods have been added by the user in a decorator? We may even do that in the `state_normalizer_class` and `state_validator_class` themselves.